File: php/testcode/xxe_025.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_xxe_external_dtd
function xxe025(BenchmarkRequest $req): BenchmarkResponse {
    $xml = $req->bodyStr();
    $dom = new DOMDocument();
    $dom->loadXML($xml, LIBXML_DTDLOAD); // vuln-code-snippet vuln-line php_xxe_external_dtd
    return BenchmarkResponse::ok($dom->saveXML());
}
// vuln-code-snippet end php_xxe_external_dtd

File: ruby/testcode/xxe_025.rb
require_relative 'shared'
require 'rexml/document'

# vuln-code-snippet start ruby_xxe_rexml_full_doc
def parse_rexml_full_doc_handler(req)
  xml = req.post('xml')
  doc = REXML::Document.new(xml) # vuln-code-snippet vuln-line ruby_xxe_rexml_full_doc
  BenchmarkResponse.json({ root: doc.root&.name })
end
# vuln-code-snippet end ruby_xxe_rexml_full_doc
